Implement D2k Death Hand & TS Cluster Missile warheads

reaperrr opened the pull request:
Replaced those "normal nukes with different label" with proper cluster warhead and weapons.

Commit:
Add FireClusterWarhead

reaperrr committed:
Add DetonationAltitude to NukePower

And RemoveMissileOnDetonation boolean.

Allows airburst, and optionally missile continuing
until it hits the ground (without a second explosion).

reaperrr committed:
Implement TS cluster missile warhead

reaperrr committed:
Implement D2k DeathHand cluster logic

GraionDilach commented:
Impressive idea on the D2k spread.

pchote commented on mods/d2k/weapons/debris.yaml:
```diff
@@ -71,6 +71,35 @@ Debris3:
 	Warhead@4Concrete: DamagesConcrete
 		Damage: 1350
 
+NukeDebris:
```
Why `NukeDebris`, and not `DeathHandCluster` or similar? The original game used `DHBOMB` and `EX_DHM` for this (#7558).

MustaphaTR commented on mods/d2k/weapons/other.yaml:
```diff
@@ -107,35 +107,18 @@ Demolish:
+		Weapon: NukeDebris
+		RandomClusterCount: 10
+		Dimensions: 3,3
+		Footprint: xxx xxx xxx
```
Looks like you did this wrong way around. TS Cluster missile is the one that is meant to fire in a 3x3 area and D2k one fires randomly on a bigger area.

reaperrr replied:
I looked at the original DH explosion, and most of the time most of the cluster would land very close to the target center.
Also, keep in mind that the weapon projectiles' inaccuracy effectively increases the max possible distance from the targeted cell.

reaperrr replied:
Checked TS too, but that only confirmed that my settings are pretty close to the original, they occasionally *did* stry quite far from the target location.

I did tweak the inaccuracy offsets (and damage) a bit, though.

MustaphaTR commented on mods/d2k/weapons/debris.yaml:
```diff
@@ -71,6 +71,35 @@ Debris3:
+	Projectile: Bullet
+		Image: 120mm
+		TrailImage: small_trail2
+		Speed: 64
```
This feels a little too slow.

reaperrr replied:
Increased.

reaperrr commented:
Updated.

- Changed warhead targeting of 'x' cells to be completely random, instead of only firing at those cells in a random order
- Fixed RandomClusterCount to actually fire all clusters even if exceeding the number of 'x' cells
- tweaked both the DeathHand and ClusterMissile, including their damage (was way too low before, especially on the DH compared to Atreides airstrike).

ltem commented:
Hi,
the PR causes the game to crash if the Air Strike ability of the Atreides is selected (clicked on it). 

<details>
<summary>Exception Log</summary>

```
OpenRA engine version {DEV_VERSION}
Dune 2000 mod version {DEV_VERSION}
on map 06be8293c93b69169596dbc26b8b01718ab1b55d (Habbanya Erg by Westwood Studios).
Date: 2019-04-12 08:48:08Z
Operating System: Linux (Unix 4.9.0.6)
Runtime Version: Mono 5.20.1.19 (tarball Thu Apr 11 09:07:22 UTC 2019) CLR 4.0.30319.42000
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object
  at OpenRA.Graphics.Animation..ctor (OpenRA.World world, System.String name, System.Func`1[TResult] facingFunc, System.Func`1[TResult] paused) [0x00041] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Graphics.Animation..ctor (OpenRA.World world, System.String name, System.Func`1[TResult] facingFunc) [0x00000] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Graphics.Animation..ctor (OpenRA.World world, System.String name) [0x00000] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Mods.Common.Traits.SelectDirectionalTarget..ctor (OpenRA.World world, System.String order, OpenRA.Mods.Common.Traits.SupportPowerManager manager, System.String cursor, System.String targetPlaceholderCursorAnimation, System.String directionArrowAnimation) [0x00068] in <ffb3ab30e4fa4867a23165aa4828e1d6>:0 
  at OpenRA.Mods.Common.Traits.AirstrikePower.SelectTarget (OpenRA.Actor self, System.String order, OpenRA.Mods.Common.Traits.SupportPowerManager manager) [0x00063] in <ffb3ab30e4fa4867a23165aa4828e1d6>:0 
  at OpenRA.Mods.Common.Traits.SupportPowerInstance.Target () [0x00038] in <ffb3ab30e4fa4867a23165aa4828e1d6>:0 
  at OpenRA.Mods.Common.Widgets.SupportPowersWidget.ClickIcon (OpenRA.Mods.Common.Widgets.SupportPowersWidget+SupportPowerIcon clicked) [0x000a1] in <ffb3ab30e4fa4867a23165aa4828e1d6>:0 
  at OpenRA.Mods.Common.Widgets.SupportPowersWidget.HandleMouseInput (OpenRA.MouseInput mi) [0x000b8] in <ffb3ab30e4fa4867a23165aa4828e1d6>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x0009d] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (OpenRA.MouseInput mi) [0x00051] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Widgets.Ui.HandleInput (OpenRA.MouseInput mi) [0x00030] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.DefaultInputHandler+<>c__DisplayClass5_0.<OnMouseInput>b__0 () [0x00000] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.DefaultInputHandler.OnMouseInput (OpenRA.MouseInput input) [0x0000d] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (OpenRA.Platforms.Default.Sdl2PlatformWindow device, OpenRA.IInputHandler inputHandler) [0x0014e] in <814709d0a3f34d14abcf0893eebcc6e3>:0 
  at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput (OpenRA.IInputHandler inputHandler) [0x00006] in <814709d0a3f34d14abcf0893eebcc6e3>:0 
  at OpenRA.Renderer.EndFrame (OpenRA.IInputHandler inputHandler) [0x0000c] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Game.RenderTick () [0x00126] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Game.Loop () [0x0010f] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Game.Run () [0x0003c] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <b83846feb5de4a5d9e4502bd4f16b18e>:0 
```
</details>

reaperrr commented:
That's a regression from the airstrike direction PR, ping @teinarss.

teinarss commented:
> 
> 
> That's a regression from the airstrike direction PR, ping @teinarss.

Will fix it this weekend!

pchote commented on OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs:
```diff
@@ -0,0 +1,126 @@
+				Weapon = weapon,
+				Facing = (map.CenterOfCell(targetCell) - target.CenterPosition).Yaw.Facing,
+
+				DamageModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IFirepowerModifier>()
```
Wouldn't it make more sense to inherit `damageModifiers` from `DoImpact`?

GraionDilach replied:
IIRC when this was originally written, DoImpact lacked that option and/or that caused crashes in weapon chaining.

pchote commented on OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs:
```diff
@@ -0,0 +1,126 @@
+				InaccuracyModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IInaccuracyModifier>()
+					.Select(a => a.GetInaccuracyModifier()).ToArray() : new int[0],
+
+				RangeModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IRangeModifier>()
```
Does this even make sense here?

GraionDilach replied:
This is coming as consistency with FSW.

pchote commented on OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs:
```diff
@@ -0,0 +1,126 @@
+				DamageModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IFirepowerModifier>()
+					.Select(a => a.GetFirepowerModifier()).ToArray() : new int[0],
+
+				InaccuracyModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IInaccuracyModifier>()
```
IMO this could be ignored to complete the set - why should the current state of the actor which may have fired this many ticks ago have any impact on the spread properties?

pchote commented on OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs:
```diff
@@ -0,0 +1,126 @@
+					firedBy.World.AddFrameEndTask(w => w.Add(projectile));
+
+				if (args.Weapon.Report != null && args.Weapon.Report.Any())
+					Game.Sound.Play(SoundType.World, args.Weapon.Report.Random(firedBy.World.SharedRandom), target.CenterPosition);
```
This would benefit from #16391

pchote commented on mods/d2k/weapons/other.yaml:
```diff
@@ -106,36 +106,51 @@ Demolish:
-		Spread: 1c0
-		Falloff: 200, 100, 50, 25, 12, 0
-		Damage: 27000 ##225 in vanilla but of course is a cluster bomb instead, so damage spread out
+		Damage: 4500
```
> tweaked both the DeathHand and ClusterMissile, including their damage (was way too low before, especially on the DH compared to Atreides airstrike).

Do you mean relative to the original game, or relative to common sense?

reaperrr replied:
Both. TS Cluster Missile was normally a guaranteed kill for at least up to Advanced Power Plants in the original, provided they were on the targeted cell.

Death Hand in the original had some inherent Missile inaccuracy, but when it hit something directly, it hit pretty hard.
Additionally, Ornithopters in our mod are more powerful than in the original, and actually feel a tiny bit more powerful than the DH even with this.
But to be honest, I'd prefer to leave any further D2k balance polishing to someone like @MustaphaTR, this was just to make the difference between the super weapons small enough that it doesn't make the DH a joke in comparison.

pchote commented on mods/d2k/weapons/other.yaml:
```diff
@@ -106,36 +106,51 @@ Demolish:
+	Warhead@4Concrete: DamagesConcrete
+		Damage: 4500
 
 CrateNuke:
```
This doesn't appear to be used at all in d2k, and if it was I don't think the cluster weapon would make sense? Best to remove it completely.

reaperrr replied:
Ok, will do.

pchote commented:
https://youtu.be/dmflM70vtZI?t=350 shows that the death hand missile doesn't actually detonate to generate the clusters, but instead spawns them before impacting the ground itself.

It also has a cool trail effect that we're still missing.

reaperrr commented:
Rebased & updated.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs:
```diff
@@ -35,6 +35,12 @@ class NukePowerInfo : SupportPowerInfo, IRulesetLoaded, Requires<BodyOrientation
 		public readonly WVec SpawnOffset = WVec.Zero;
 
+		[Desc("Altitude offset from the target position at which the warhead should detonate.")]
+		public readonly WDist PreDetonationAltitude = WDist.Zero;
```
I preferred the idea of having separate `AirburstAltitude` and `PreDetonationAltitude`, but if you want to have them combined then this should be kept as `AirbustAltitude` and `RemoveMissileOnAirburst`.

The whole motivation for the "PreDetonation" name was that it would never detonate/remove the missile.

pchote replied:
On second thought, the better option is to just remove the "Pre". `DetonationAltitude` and `RemoveMissileOnDetonation` sound much better than Airburst, which we only considered because of the RA2 history.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs:
```diff
@@ -35,6 +35,12 @@ class NukePowerInfo : SupportPowerInfo, IRulesetLoaded, Requires<BodyOrientation
+		[Desc("Altitude offset from the target position at which the warhead should detonate.")]
+		public readonly WDist DetonationAltitude = WDist.Zero;
+
+		[Desc("Custom palette is a player palette BaseName.")]
```
Incorrect description.

pchote commented on OpenRA.Mods.Common/Projectiles/NukeLaunch.cs:
```diff
@@ -116,23 +125,33 @@ public void Tick(World world)
 				trailTicks = trailInterval;
 			}
 
+			if (isDescending && preDetonates && pos.Z <= detonationAltitude.Length)
```
These two cases can be combined, and the `preDetonates` flag removed:
```csharp
if (ticks == impactDelay || (isDescending && pos.Z <= detonationAltitude.Length))
	Explode(world, ticks == impactDelay || removeOnDetonation);
```

pchote replied:
This also doesn't seem to be accounting for terrain height in TS - cluster missiles launched on high ground impact the ground then send clusters horizontally.

reaperrr replied:
Should be DistanceAboveTerrain instead of pos.Z, I guess.

reaperrr commented:
Updated. Did a quick test and DeathHand still works, while TS cluster missile now correctly considers terrain height.

reaperrr commented:
Getting the updated DeathHand into the playtest would be nice, adding to milestone.

teinarss commented:
Looks like the launch of the cluster is a bit off? Not sure how it was in the original.

![cluster](https://user-images.githubusercontent.com/19173272/57397815-ccd11680-71cd-11e9-90b2-85f0a2acf94a.gif)

pchote commented:
Also looks like its missing a launch delay to let the doors open before firing.

Needs a rebase.

reaperrr committed:
Fix launch delay and offset of TS ClusterMissile

reaperrr commented:
Rebased, fixed offset and launch delay.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/b80c09b2331fc1a4f2b5e158a7ae209f286f4c44
